Packet++/src/DoIpLayer.cpp

```diff
@@ -0,0 +1,1381 @@
+		case DoIpPayloadTypes::ENTITY_STATUS_REQUEST:
+			return new DoIpEntityStatusRequest(data, dataLen, prevLayer, packet);
+		default:
+			return nullptr;
+		}
+	}
+	std::string DoIpLayer::getProtocolVersionAsStr() const
+	{
+		auto it = DoIpEnumToStringProtocolVersion.find(getProtocolVersion());
+		return it->second;
+	}
+
+	void DoIpLayer::setProtocolVersion(DoIpProtocolVersion version)
+	{
+		getDoIpHeader()->protocolVersion = static_cast<uint8_t>(version);
+	}
+
+	void DoIpLayer::setProtocolVersion(uint8_t version)
+	{
+		getDoIpHeader()->protocolVersion = version;
+	}
+
+	uint8_t DoIpLayer::getInvertProtocolVersion() const
+	{
+		return getDoIpHeader()->invertProtocolVersion;
+	}
+
+	void DoIpLayer::setInvertProtocolVersion(uint8_t iVersion)
+	{
+		getDoIpHeader()->invertProtocolVersion = iVersion;
+	}
+
+	void DoIpLayer::setPayloadType(DoIpPayloadTypes type)
+	{
+		getDoIpHeader()->payloadType = htobe16(static_cast<uint16_t>(type));
+	}
+
+	std::string DoIpLayer::getPayloadTypeAsStr() const
+	{
+		auto it = DoIpEnumToStringPayloadType.find(getPayloadType());
+		return it->second;
+	}
+
+	uint32_t DoIpLayer::getPayloadLength() const
+	{
+		return htobe32(getDoIpHeader()->payloadLength);
+	}
+
+	void DoIpLayer::setPayloadLength(uint32_t payloadLength)
+	{
+		getDoIpHeader()->payloadLength = be32toh(payloadLength);
+	}
+
+	std::string DoIpLayer::toString() const
+	{
+		std::ostringstream oss;
+		DoIpPayloadTypes type = getPayloadType();
+
+		oss << "DoIP Layer, " << getPayloadTypeAsStr() << " (0x" << std::hex << std::setw(4) << std::setfill('0')
+		    << static_cast<uint16_t>(type) << ")";
+
+		return oss.str();
+	}
+
+	void DoIpLayer::setHeaderFields(DoIpProtocolVersion version, DoIpPayloadTypes type, uint32_t length)
+	{
+		setProtocolVersion(version);
+		setInvertProtocolVersion(~(static_cast<uint8_t>(version)));
+		setPayloadType(type);
+		setPayloadLength(length);
+	}
+
+	void DoIpLayer::parseNextLayer()
+	{
+		if (getPayloadType() == DoIpPayloadTypes::DIAGNOSTIC_MESSAGE_TYPE)
+		{
+			size_t headerLen = getHeaderLen();
+
+			if (m_DataLen <= headerLen)
+			{
+				return;
+			}
+
+			uint8_t* payload = m_Data + headerLen;
+			size_t payloadLen = m_DataLen - headerLen;
+
+			constructNextLayer<PayloadLayer>(payload, payloadLen, m_Packet);
+		}
+	}
+
+	//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~|
+	// DoIpRoutingActivationRequest|
+	//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~|
+	DoIpRoutingActivationRequest::DoIpRoutingActivationRequest(uint8_t* data, size_t dataLen, Layer* prevLayer,
+	                                                           Packet* packet)
+	    : DoIpLayer(data, dataLen, prevLayer, packet)
+	{
+		if (dataLen < FIXED_LEN || dataLen > OPT_LEN)
+			throw std::runtime_error("DoIpRoutingActivationRequest: Invalid payload length!");
+
+		if (dataLen > FIXED_LEN && dataLen < OPT_LEN)
+			throw std::runtime_error("DoIpRoutingActivationRequest: Invalid OEM field length!");
+
+		if (dataLen == OPT_LEN)
+		{
+			_hasReservedOem = true;
```

Comment:
We probably don't need this private member because it's very easy to calculate it on the fly:
```cpp
bool DoIpRoutingActivationRequest::hasReservedOem() const
{
	return m_DataLen == OPT_LEN;
}
```